library/core/src/iter/adapters/intersperse.rs

```diff
@@ -40,37 +40,143 @@ where
         }
     }
 
-    fn fold<B, F>(mut self, init: B, mut f: F) -> B
+    fn fold<B, F>(self, init: B, f: F) -> B
     where
         Self: Sized,
         F: FnMut(B, Self::Item) -> B,
     {
-        let mut accum = init;
+        let separator = self.separator;
+        intersperse_fold(self.iter, init, f, move || separator.clone(), self.needs_sep)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        intersperse_size_hint(&self.iter, self.needs_sep)
+    }
+}
+
+/// An iterator adapter that places a separator between all elements.
+#[unstable(feature = "iter_intersperse", reason = "recently added", issue = "79524")]
+pub struct IntersperseWith<I, G>
+where
+    I: Iterator,
+{
+    separator: G,
+    iter: Peekable<I>,
+    needs_sep: bool,
+}
+
+#[unstable(feature = "iter_intersperse", reason = "recently added", issue = "79524")]
+impl<I, G> crate::fmt::Debug for IntersperseWith<I, G>
+where
```

Comment:
What's the reason for not using `#[derive(Debug)]`? Same question for `Clone`.

Reply:
The problem is that `derive` generates code which in turn causes a compile-error unless bounds for `Debug` and `Clone` are placed on the type-parameters in the `struct`-definition(!) itself. This is obviously not what we want, because we would not be able to construct an `Intersperse`/ `IntersperseWith` for inner types that are not `Debug` and `Clone`.
I asked [here](https://internals.rust-lang.org/t/misplaced-trait-bounds-using-derive-in-stdlib/13759) about it and removed the `FIXME` as advised.

Reply:
Ah, I was confused because `Intersperse` does use `#[derive(Debug, Clone)]`. But the subtle and important difference is that that one has an `I::Item` member, which `#[derive]` understands.